docs/Practices/hub/hub.md

```diff
@@ -0,0 +1,118 @@
+
+For the purpose of this document, conventions should be seen as more specific guidelines or extensions to the guidelines.
+
+Naming:
+- The entrypoint function for a command or query should start with `command` or `query`.
+- The entrypoint function for a command or query should be in a file called `command.ts` or `query.ts` OR in a region called `Command` or `Query`.
+- Domain code should either be in files named appropriate to the model/process/rule/etc. within a `domain` directory, or be in a `domain.ts` file, or be in a region called `Domain`.
+- Legacy request handlers that are related to a bounded context and are expected to be maintained/iterated should be put in a `legacy-handlers` directory within the appropriate bounded context.
+- Integration code for 3rd party services should be in a `Data` directory.
+- Repository code (code that interacts with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.
+- (Request or Event) Handlers exposing a command should be in a file called `api.ts`, a region called `API`, or in files called `request.ts` or `event.ts` in a directory called `handlers`.
+- Integration code for bounded contexts in a different project should be in a `dependent-bcs` directory with sub directories matching tha appropriate bounded context. *As of Jan 24 2023 this is not being utilized. Defining this early to reduce decision overhead when this occurs.*
+- CRUD type functions that create new records in a single tab, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.
+
+*[Hub Directory Structure Example](examples.md#hub-directory-structure)*
+
+Tests:
+- Test files should sit close to the code they test. *Test files that test code in disparate files should be broken up into multiple test files itself.*
+
+SQL:
+- Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to it's consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*
+
+Routing:
+- Handlers should be wired up to routes in files within a directory called `routes` that is at the bounded context root. **
+
+## The Future
+- In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in one of the schemas in our postgres instance and utilizing a single shared layer among all consumers within Hub might be the most pragmatic approach.*
```

Comment:
YES, maybe some shapes for basic DB objects. Or if we can use a DB middleware that can returned typed objects (is that even a thing?)

Reply:
To be very clear in the communication of this point. I'll add a distinction to this text in the document as well as here in comment. I do mean single shared **repository** -layer- model.

There's plenty of tools out there that will returned typed objects. I've used TypeOrm for quite a while and I like it. While it does require some configuration, all the options make sense, it's really mature, well maintained, and not overbearing.

The key things I want in a DB tool is:
- Validate during process startup/making a connection that code defined shapes (type/interface/class) are valid representations of their DB table counterparts and throw/return an error if this is not the case.
- Return to you typed objects when interacting with it's API
- Able to execute custom SQL and provide a generic argument for return query shape.

Reply:
I wonder if we could add [Prisma](https://www.prisma.io/typescript) as the DB query/update tool. Maybe down the road we could make it the full ORM. At least it would give us the types, and have the option to run straight SQL too.

Reply:
Prisma works tool. Tool choice doesn't matter much to me as much as typesafety and flexibility.